<a href="https://colab.research.google.com/github/ljcotica/DataScience/blob/main/Importing_Datasets_with_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Aquisição dos dados



Existem vários formatos para um conjunto de dados, .csv, .json, .xlsx etc. O conjunto de dados pode ser armazenado em diferentes lugares, em sua máquina local ou às vezes online.
Nesta exemplo, carregaremos um conjunto de dados em nosso Jupyter Notebook.
No nosso caso, o Automobile Dataset é uma fonte online e está no formato CSV (valores separados por vírgula). Vamos usar este conjunto de dados como um exemplo para praticar a leitura de dados.

*   fonte de dados: https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
*   tipo de dados: csv

A Biblioteca Pandas é uma ferramenta útil que nos permite ler vários conjuntos de dados em um quadro de dados.

Adicionando a biblioteca:

In [1]:
# Importa as biblioteca Pandas e Numpy
import pandas as pd
import numpy as np

# Lendo os dados

Usamos a função *pandas.read_csv()* para ler o arquivo csv. Entre colchetes, colocamos o caminho do arquivo junto com uma aspa, para que o Pandas leia o arquivo em um quadro de dados daquele endereço. O caminho do arquivo pode ser um URL ou o endereço do arquivo local.
Como os dados não incluem cabeçalhos, podemos adicionar um argumento *headers = None* dentro do método* read_csv()*, para que o Pandas não defina automaticamente a primeira linha como cabeçalho.
Você também pode atribuir o conjunto de dados a qualquer variável que criar.

In [2]:
# Lê o arquivo online que está na URL definida abaixo e atribui os valores a uma variável chamada "df"
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data"
df = pd.read_csv(url, header = None)

Depois de ler o conjunto de dados, podemos usar o método *dataframe.head(n)* para verificar as primeiras n linhas do dataframe; onde n é um número inteiro. Ao contrário de *dataframe.head(n)*, *dataframe.tail(n)* mostrará as n linhas inferiores do dataframe.

In [3]:
# mostra as 5 primeiras linhas usando o método dataframe.head()
print("As 5 primeiras linhas do dataframe") 
df.head(5)

As 5 primeiras linhas do dataframe


0    1            2    3    4     5   ...    20   21    22  23  24     25
0   3    ?  alfa-romero  gas  std   two  ...   9.0  111  5000  21  27  13495
1   3    ?  alfa-romero  gas  std   two  ...   9.0  111  5000  21  27  16500
2   1    ?  alfa-romero  gas  std   two  ...   9.0  154  5000  19  26  16500
3   2  164         audi  gas  std  four  ...  10.0  102  5500  24  30  13950
4   2  164         audi  gas  std  four  ...   8.0  115  5500  18  22  17450

[5 rows x 26 columns]

In [4]:
# mostra as 5 últimas linhas usando o método dataframe.head()
print("As 5 últimas linhas do dataframe") 
df.tail(5)

As 5 últimas linhas do dataframe


0   1      2       3      4     5   ...    20   21    22  23  24     25
200  -1  95  volvo     gas    std  four  ...   9.5  114  5400  23  28  16845
201  -1  95  volvo     gas  turbo  four  ...   8.7  160  5300  19  25  19045
202  -1  95  volvo     gas    std  four  ...   8.8  134  5500  18  23  21485
203  -1  95  volvo  diesel  turbo  four  ...  23.0  106  4800  26  27  22470
204  -1  95  volvo     gas  turbo  four  ...   9.5  114  5400  19  25  22625

[5 rows x 26 columns]

# Adicionando cabeçalho

Dê uma olhada em nosso conjunto de dados; O Pandas define automaticamente o cabeçalho como um inteiro começando por 0.

Para descrever melhor nossos dados, podemos apresentar um cabeçalho, essa informação está disponível em: https://archive.ics.uci.edu/ml/datasets/Automobile

Portanto, temos que adicionar cabeçalhos manualmente.

Em primeiro lugar, criamos uma lista de "cabeçalhos" que inclui todos os nomes das colunas em ordem. Em seguida, usamos dataframe.columns = headers para substituir os cabeçalhos pela lista que criamos.



In [5]:
# cria uma lista de cabeçalho
headers=["symboling","normalized-losses","make","fuel-type","aspiration","num-of-doors","body-style","drive-wheels","engine-location","wheel-base","length","width","height","curb-weight","engine-type","num-of-cylinders","engine-size","fuel-system","bore","stroke","compression-ratio","horsepower","peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", headers)

headers
 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


Substituímos os cabeçalhos e verificamos novamente nosso dataframe

In [6]:
df.columns = headers
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

Descrição das colunas:

1. symboling:                -3, -2, -1, 0, 1, 2, 3.
2. normalized-losses:        continuous from 65 to 256.
3. make:                     alfa-romero, audi, bmw, chevrolet, dodge, honda,isuzu, jaguar, mazda, mercedes-benz, mercury, mitsubishi, nissan, peugot, plymouth, porsche,renault, saab, subaru, toyota, volkswagen, volvo
4. fuel-type:                diesel, gas.
5. aspiration:               std, turbo.
6. num-of-doors:             four, two.
7. body-style:               hardtop, wagon, sedan, hatchback, convertible.
8. drive-wheels:             4wd, fwd, rwd.
9. engine-location:          front, rear.
10. wheel-base:               continuous from 86.6 120.9.
11. length:                   continuous from 141.1 to 208.1.
12. width:                    continuous from 60.3 to 72.3.
13. height:                   continuous from 47.8 to 59.8.
14. curb-weight:              continuous from 1488 to 4066.
15. engine-type:              dohc, dohcv, l, ohc, ohcf, ohcv, rotor.
16. num-of-cylinders:         eight, five, four, six, three, twelve, two.
17. engine-size:              continuous from 61 to 326.
18. fuel-system:              1bbl, 2bbl, 4bbl, idi, mfi, mpfi, spdi, spfi.
19. bore:                     continuous from 2.54 to 3.94.
20. stroke:                   continuous from 2.07 to 4.17.
21. compression-ratio:        continuous from 7 to 23.
22. horsepower:               continuous from 48 to 288.
23. peak-rpm:                 continuous from 4150 to 6600.
24. city-mpg:                 continuous from 13 to 49.
25. highway-mpg:              continuous from 16 to 54.
26. price:                    continuous from 5118 to 45400.

Precisamos substituir o *"?"* símbolo com NaN para que o *dropna()* possa remover os valores ausentes

In [7]:
df1=df.replace('?',np.NaN)

Podemos descartar os valores ausentes da coluna "preço" da seguinte maneira

In [8]:
df=df1.dropna(subset=["price"], axis=0)
df.head(20)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0           3               NaN  alfa-romero  ...       21          27  13495
1           3               NaN  alfa-romero  ...       21          27  16500
2           1               NaN  alfa-romero  ...       19          26  16500
3           2               164         audi  ...       24          30  13950
4           2               164         audi  ...       18          22  17450
5           2               NaN         audi  ...       19          25  15250
6           1               158         audi  ...       19          25  17710
7           1               NaN         audi  ...       19          25  18920
8           1               158         audi  ...       17          20  23875
10          2               192          bmw  ...       23          29  16430
11          0               192          bmw  ...       23          29  16925
12          0               188          bmw  ...       21          28  20970
13          0               188          bmw  ...       21          28  21105
14          1               NaN          bmw  ...       20          25  24565
15          0               NaN          bmw  ...       16          22  30760
16          0               NaN          bmw  ...       16          22  41315
17          0               NaN          bmw  ...       15          20  36880
18          2               121    chevrolet  ...       47          53   5151
19          1                98    chevrolet  ...       38          43   6295
20          0                81    chevrolet  ...       38          43   6575

[20 rows x 26 columns]

Agora, lemos com sucesso o conjunto de dados e adicionamos os cabeçalhos corretos.

# Salvando o dataset

Da mesma forma, o Pandas nos permite salvar o conjunto de dados em csv usando o método *dataframe.to_csv()*, você pode adicionar o caminho e o nome do arquivo junto com as aspas entre colchetes.

Por exemplo, se você deseja salvar o dataframe *df* como *automovel.csv* em sua máquina local, você pode usar a sintaxe abaixo:

In [9]:
df.to_csv("automovel.csv", index=False)

Também podemos ler e salvar outros formatos de arquivo, podemos usar funções semelhantes a *pd.read_csv()* e *df.to_csv()* para outros formatos de dados, as funções estão listadas na tabela a seguir:

# Ler/Salvar outros formatos

<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |

# Insight básico do dataset

Depois de ler os dados no dataframe do Pandas, é hora de explorarmos o conjunto de dados.<br>
Existem várias maneiras de obter insights essenciais dos dados para nos ajudar a entender melhor nosso conjunto de dados.


## Tipos de dados

Os dados têm vários tipos. <br>
Os principais tipos armazenados nos dataframes do Pandas são object, float, int, bool e datetime64. Para conhecer melhor cada atributo, é sempre bom sabermos o tipo de dado de cada coluna. Em Pandas:

In [10]:
# verifica o tipo de dados do dataframe "df"
print(df.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


Retorna uma série com o tipo de dados de cada coluna. <br>
Como resultado, conforme mostrado acima, é claro ver que o tipo de dados de "symboling" e "curb-weight" são int64, "normalized-losses" é object e "wheel-base" é float64, etc. <br>

Esses tipos de dados podem ser alterados.

# Descrição

Se quisermos obter um resumo estatístico de cada coluna, como contagem, valor médio da coluna, desvio padrão da coluna, etc. Usamos o método *describe*:

*dataframe.describe()*

Este método fornecerá várias estatísticas resumidas, excluindo os valores NaN (not a number). 

In [11]:
df.describe()

symboling  wheel-base  ...    city-mpg  highway-mpg
count  201.000000  201.000000  ...  201.000000   201.000000
mean     0.840796   98.797015  ...   25.179104    30.686567
std      1.254802    6.066366  ...    6.423220     6.815150
min     -2.000000   86.600000  ...   13.000000    16.000000
25%      0.000000   94.500000  ...   19.000000    25.000000
50%      1.000000   97.000000  ...   24.000000    30.000000
75%      2.000000  102.400000  ...   30.000000    34.000000
max      3.000000  120.900000  ...   49.000000    54.000000

[8 rows x 10 columns]

Isso mostra o resumo estatístico de todas as colunas de tipo numérico (int, float).<br>
Por exemplo, o atributo "symboling" tem 205 contagens, o valor médio desta coluna é 0.83, o desvio padrão é 1.25, o valor mínimo é -2, 25º percentil é 0, 50º percentil é 1, 75º percentil é 2 e o valor máximo é 3.<br>
No entanto, se também quisermos verificar todas as colunas, incluindo aquelas que são do tipo objeto.
Você pode adicionar um argumento *include = "all"* dentro do colchete. Vamos tentar novamente.

In [12]:
# descreve todas as colunas no"df" 
df.describe(include = "all")

symboling normalized-losses    make  ...    city-mpg highway-mpg price
count   201.000000               164     201  ...  201.000000  201.000000   201
unique         NaN                51      22  ...         NaN         NaN   186
top            NaN               161  toyota  ...         NaN         NaN  8921
freq           NaN                11      32  ...         NaN         NaN     2
mean      0.840796               NaN     NaN  ...   25.179104   30.686567   NaN
std       1.254802               NaN     NaN  ...    6.423220    6.815150   NaN
min      -2.000000               NaN     NaN  ...   13.000000   16.000000   NaN
25%       0.000000               NaN     NaN  ...   19.000000   25.000000   NaN
50%       1.000000               NaN     NaN  ...   24.000000   30.000000   NaN
75%       2.000000               NaN     NaN  ...   30.000000   34.000000   NaN
max       3.000000               NaN     NaN  ...   49.000000   54.000000   NaN

[11 rows x 26 columns]

Agora, ele fornece o resumo estatístico de todas as colunas, incluindo atributos digitados por objeto.<br>
Agora podemos ver quantos valores únicos, que é o valor superior e a frequência do valor superior nas colunas digitadas por objeto.<br>
Alguns valores na tabela acima são mostrados como "NaN", porque esses números não estão disponíveis para um tipo de coluna específico.

Você pode selecionar as colunas de um quadro de dados, indicando o nome de cada coluna, por exemplo, você pode selecionar as três colunas da seguinte forma:
<br><br>
dataframe [['coluna 1', coluna 2 ',' coluna 3 ']]
<br><br>
Onde "coluna" é o nome da coluna, você pode aplicar o método *".describe ()"* para obter as estatísticas dessas colunas da seguinte maneira:
<br><br>
dataframe [['coluna 1', coluna 2 ',' coluna 3 ']] .describe ()


In [13]:
#Aplicando o método ".describe()" às colunas 'length' e 'compression-ratio'.
df[['length','compression-ratio']].describe()

length  compression-ratio
count  201.000000         201.000000
mean   174.200995          10.164279
std     12.322175           4.004965
min    141.100000           7.000000
25%    166.800000           8.600000
50%    173.200000           9.000000
75%    183.500000           9.400000
max    208.100000          23.000000

## Informações

Outro método que você pode usar para verificar seu conjunto de dados é:

*dataframe.info()*

Ele fornece um resumo conciso do seu Dataframe.
<br>
Este método imprime informações sobre um Dataframe incluindo o índice dtype e colunas, valores não nulos e uso de memória.

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       199 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non